app/backend/approaches/chatreadretrieveread.py

```diff
@@ -127,61 +129,60 @@ async def run_until_final_call(
-        user_query_request = "Generate search query for: " + original_user_query
 
-        tools: List[ChatCompletionToolParam] = [
-            {
-                "type": "function",
-                "function": {
-                    "name": "search_sources",
-                    "description": "Retrieve sources from the Azure AI Search index",
-                    "parameters": {
-                        "type": "object",
-                        "properties": {
-                            "search_query": {
-                                "type": "string",
-                                "description": "Query string to retrieve documents from azure search eg: 'Health care plan'",
-                            }
-                        },
-                        "required": ["search_query"],
-                    },
-                },
-            }
-        ]
+        # Process results
 
-        # STEP 1: Generate an optimized keyword search query based on the chat history and the last question
-        query_response_token_limit = 100
-        query_messages = build_messages(
-            model=self.chatgpt_model,
-            system_prompt=self.query_prompt_template,
-            tools=tools,
+        # Load the Prompty objects for AI Search query and chat answer generation.
+        prompty_path = SUPPORTED_MODELS.get(current_model)
+        if prompty_path:
+            chat_prompty = Prompty.load(source=prompty_path / "chat.prompty")
+            query_prompty = Prompty.load(source=prompty_path / "query.prompty")
+        else:
+            raise ValueError(f"Model {current_model} is not supported. Please create a template for this model.")
+        # Shorten the past messages if needed
+        question_token_limit = chat_prompty._model.configuration.get(
+            "messages_length_limit", 4000
+        ) - chat_prompty._model.parameters.get("max_tokens", 1024)
+
+        past_messages = shorten_past_messages(
+            model=current_model,
+            model_type=query_prompty._model.configuration["type"],
+            system_message=self.query_prompt_template,
+            max_tokens=question_token_limit,
+            tools=query_prompty._model.parameters.get("tools"),
+            new_user_content=original_user_query,
             few_shots=self.query_prompt_few_shots,
             past_messages=messages[:-1],
-            new_user_content=user_query_request,
-            max_tokens=self.chatgpt_token_limit - query_response_token_limit,
         )
-
+        # STEP 1: Generate an optimized keyword search query based on the chat history and the last question
+        query_messages = query_prompty.render(
+            system_message=self.query_prompt_template,
+            question=original_user_query,
+            few_shots=self.query_prompt_few_shots,
+            past_messages=past_messages,
+        )
+        query_messages = ast.literal_eval(query_messages)
```

Comment:
Why are we evaluating the parsed messages? I understand we get a `str` representation of a list of messages. Is this just to convert it back to a Python `list`?

Reply:
> Why are we evaluating the parsed messages? I understand we get a `str` representation of a list of messages. Is this just to convert it back to a Python `list`?

Yep, that is solely to convert to a dictionary as for some reason Prompty.render returns a string.